### PR TITLE
Differentiate between URI and SSL settings for in-cluster vs. submission

### PR DESCRIPTION
**Pull request by mccheah**
Allows users to provide separate settings for contacting the resource staging server from within the init-container vs. from the submission client. One example use case for this is that the resource staging server might be running inside the cluster, and exposed to the submission client through an Ingress or some other front door URI. In this case, the resource staging server should be accessed from the submission client via the Ingress URI, but the init-containers that are running inside the cluster should access the resource staging server through the pod IP or service IP. The certificate that has to be presented when contacting over the external URI may also differ from the certificate that needs to be used when contacting through the internal URI.

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala [hunk 1]
```diff
@@ -370,47 +372,98 @@ package object config extends Logging {
       .createOptional
   private[spark] val RESOURCE_STAGING_SERVER_CLIENT_CERT_PEM =
     ConfigBuilder(s"spark.ssl.$RESOURCE_STAGING_SERVER_SSL_NAMESPACE.clientCertPem")
-      .doc("Certificate PEM file to use when the client contacts the resource staging server.")
+      .doc("Certificate PEM file to use when the client contacts the resource staging server." +
+        " This must strictly be a path to a file on the submitting machine's disk.")
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:376]: with a scheme like `file://` or without?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala [hunk 2]
```diff
@@ -370,47 +372,98 @@ package object config extends Logging {
...
+      .stringConf
+      .createOptional
+  private[spark] val RESOURCE_STAGING_SERVER_INTERNAL_CLIENT_CERT_PEM =
+    ConfigBuilder(s"spark.ssl.$RESOURCE_STAGING_SERVER_INTERNAL_SSL_NAMESPACE.clientCertPem")
+      .doc("Certificate PEM file to use when the init-container contacts the resource staging" +
+        " server. If this is not provided, it defaults to the value of" +
+        " spark.ssl.kubernetes.resourceStagingServer.clientCertPem. This can be a URI with" +
+        " a scheme of local:// which denotes that the file is pre-mounted on the init-container's" +
+        " disk. A uri without a scheme or a scheme of file will result in this file being" +
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: file -> file://

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala [hunk 3]
```diff
@@ -370,47 +372,98 @@ package object config extends Logging {
...
+      .doc("Whether or not to use SSL when communicating with the resource staging server from" +
+        " the init-container. If this is not provided, defaults to" +
+        " the value of spark.ssl.kubernetes.resourceStagingServer.enabled")
       .booleanConf
       .createOptional
   private[spark] val RESOURCE_STAGING_SERVER_TRUSTSTORE_FILE =
     ConfigBuilder(s"spark.ssl.$RESOURCE_STAGING_SERVER_SSL_NAMESPACE.trustStore")
-      .doc("File containing the trustStore to communicate with the Kubernetes dependency server.")
+      .doc("File containing the trustStore to communicate with the Kubernetes dependency server." +
+        " This must strictly be a path on the submitting machine's disk.")
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:416]: what scheme?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala [hunk 4]
```diff
@@ -47,10 +48,49 @@ private[spark] class DriverInitContainerComponentsProviderImpl(
...
 
   private val maybeResourceStagingServerUri = sparkConf.get(RESOURCE_STAGING_SERVER_URI)
+  private val maybeResourceStagingServerInternalUri =
+      sparkConf.get(RESOURCE_STAGING_SERVER_INTERNAL_URI)
+  private val maybeResourceStagingServerInternalTrustStore =
+      sparkConf.get(RESOURCE_STAGING_SERVER_INTERNAL_TRUSTSTORE_FILE)
+          .orElse(sparkConf.get(RESOURCE_STAGING_SERVER_TRUSTSTORE_FILE))
+  private val maybeResourceStagingServerInternalTrustStorePassword =
+      sparkConf.get(RESOURCE_STAGING_SERVER_INTERNAL_TRUSTSTORE_PASSWORD)
+        .orElse(sparkConf.get(RESOURCE_STAGING_SERVER_TRUSTSTORE_PASSWORD))
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala]: this .orElse isn't aligned with the others

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala [hunk 5]
```diff
@@ -47,10 +48,49 @@ private[spark] class DriverInitContainerComponentsProviderImpl(
...
+          .orElse(sparkConf.get(RESOURCE_STAGING_SERVER_TRUSTSTORE_TYPE))
+  private val maybeResourceStagingServerInternalClientCert =
+      sparkConf.get(RESOURCE_STAGING_SERVER_INTERNAL_CLIENT_CERT_PEM)
+          .orElse(sparkConf.get(RESOURCE_STAGING_SERVER_CLIENT_CERT_PEM))
+  private val resourceStagingServerInternalSslEnabled =
+      sparkConf.get(RESOURCE_STAGING_SERVER_INTERNAL_SSL_ENABLED)
+          .orElse(sparkConf.get(RESOURCE_STAGING_SERVER_SSL_ENABLED))
+          .getOrElse(false)
+
+  require(
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala]: can you not use the requireNand from elsewhere here?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala [hunk 6]
```diff
@@ -47,10 +48,49 @@ private[spark] class DriverInitContainerComponentsProviderImpl(
...
+        maybeResourceStagingServerInternalTrustStore.isDefined),
+    "Cannot provide both a certificate file and a trustStore file for init-containers to" +
+      " use for contacting the resource staging server over TLS.")
+  require(maybeResourceStagingServerInternalTrustStore.forall { trustStore =>
+    Option(Utils.resolveURI(trustStore).getScheme).getOrElse("file") match {
+      case "file" | "local" => true
+      case _ => false
+    }
+  }, "TrustStore URI used for contacting the resource staging server from init containers must" +
+    " have no scheme, or scheme file, or scheme local.")
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala]: file:// and local://

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala [hunk 7]
```diff
@@ -47,10 +48,49 @@ private[spark] class DriverInitContainerComponentsProviderImpl(
...
+  require(maybeResourceStagingServerInternalClientCert.forall { trustStore =>
+    Option(Utils.resolveURI(trustStore).getScheme).getOrElse("file") match {
+      case "file" | "local" => true
+      case _ => false
+    }
+  }, "Client cert file URI used for contacting the resource staging server from init containers" +
+    " must have no scheme, or scheme file, or scheme local.")
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala]: file:// and local://

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala [hunk 8]
```diff
@@ -322,7 +322,8 @@ private[spark] class KubernetesClusterSchedulerBackend(
             .addToRequests("cpu", executorCpuQuantity)
             .addToLimits("cpu", executorCpuQuantity)
           .endResources()
-          .withEnv(requiredEnv.asJava)
+          .addAllToEnv(requiredEnv.asJava)
+          .addToEnv(executorExtraClasspathEnv.toSeq: _*)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:326]: was this not being used before?  I'm surprised checkstyle or something didn't alert that there was an unused identifier

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/InitContainerResourceStagingServerSslOptionsProvider.scala [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+package org.apache.spark.deploy.kubernetes.submit.v2
+
+/**
+ * Provides the SSL configurations that should be used when the init-container contacts the
+ * resource staging server. This is responsible for checking if the options for the init-
+ * container should be different from the options used for the submission client.
+ */
+trait InitContainerResourceStagingServerSslOptionsProvider {
+
+  def getSslOptions(): InitContainerResourceStagingServerSslOptions
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/InitContainerResourceStagingServerSslOptionsProvider.scala]: where is this called from?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala (hunk 1 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:376]: "Strictly a path" I think implies no scheme.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala (hunk 3 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:416]: "Strictly be a path" denotes no scheme.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/InitContainerResourceStagingServerSslOptionsProvider.scala (hunk 9 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/InitContainerResourceStagingServerSslOptionsProvider.scala]: I think this was an original idea I had but abandoned it later, forgot to remove the class.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala (hunk 8 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:326]: Yeah unfortunately this slipped under the radar somehow. Are the SBT tests running now? Maybe SBT would have caught it and Maven doesn't?

Comment (ash211) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:326]: they're not running right now, but I can temporarily re-enable, you push a commit, and we watch them run to see if they would've caught this

Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:326]: No need specifically but I suppose it's all the more reason to look into getting our SBT builds online.

**Comment by mccheah** [comment]
Requires rebasing onto allow-client-cert-pem

**Comment by mccheah** [comment]
Finished rebasing here for now.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/SubmittedDependencyInitContainerConfigPlugin.scala [hunk 10]
```diff
@@ -34,36 +35,62 @@ private[spark] trait SubmittedDependencyInitContainerConfigPlugin {
...
+  private def resolveSecretPath(
+      maybeUri: Option[String],
+      secretKey: String,
+      configEntry: OptionalConfigEntry[String],
+      uriType: String): Map[String, String] = {
+    maybeUri.map(Utils.resolveURI).map { uri =>
+      val resolvedPath = Option(uri.getScheme).getOrElse("file") match {
+        case "file" => s"$secretsVolumeMountPath/$secretKey"
+        case "local" => uri.getPath
+        case invalid => throw new SparkException(s"$uriType has invalid scheme $invalid must be" +
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/SubmittedDependencyInitContainerConfigPlugin.scala:90]: maybe in a fixup commit after this merges can add "has invalid scheme for secret" to give end user just a bit more context

**Review by ash211 — APPROVED** [review]
LGTM, planning to merge when this integration test passes

**Pull request merged**